### PR TITLE
Move CI checks to Github Actions

### LINKED ISSUE #38: CI is no longer running
It looks like we have some config for running CI at https://travis-ci.org, which is no longer running. We should convert to github actions, and get rid of the old Travis config (and fix the broken badge in the README)

### PR DESCRIPTION
**Pull request by H-Shay**
Fixes #38. Moved CI checks to Github Actions, removed old Travis configs, and reformatted some code that was failing `black` invocation.

### CODE REVIEW

**Comment by H-Shay** [comment]
One thing I want to note is that the unit test invocations are failing because they are set to fail if coverage is below 100%. The test coverage is currently at 85% due to the fact that either lines 26-30 OR lines 31-35 will be tested depending on which `frozendict` implementation (c-extension or no c-extension) the user is running, thus preventing us from having 100% coverage. I will see what I can do about that in another PR but I wanted to mention it.

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+        with:
+          python-version: "3.7"
+      - run: python -m pip install tox
+      - run: tox -e py37
+
+  run-unit-tests-38:
```
Comment (DMRobertson) [inline .github/workflows/tests.yaml:65]: I think we should also run these tests on 3.9 and the recently released 3.10 (if setup-python supports it)

**Review by DMRobertson — CHANGES_REQUESTED** [review]
We could reduce some of the deduplication here using a [`strategy.matrix`](https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix). https://github.com/matrix-org/synapse/blob/develop/.github/workflows/tests.yml is an example.

I'd be tempted to have one matrix for linting (black and pep8); one for the two packaging jobs, and one for the unit tests.

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (DMRobertson) [reply .github/workflows/tests.yaml:65]: Looks like 3.10 is supported as of about three hours ago! https://github.com/actions/python-versions/releases/tag/3.10.0-117470

**Comment by DMRobertson** [comment]
> One thing I want to note is that the unit test invocations are failing because they are set to fail if coverage is below 100%. The test coverage is currently at 85% due to the fact that either lines 26-30 OR lines 31-35 will be tested depending on which `frozendict` implementation (c-extension or no c-extension) the user is running, thus preventing us from having 100% coverage. I will see what I can do about that in another PR but I wanted to mention it.

I think "coverage combine" might be what we want here. Not familiar with the details, but maybe https://github.com/matrix-org/synapse/blob/6fcc3e0bc81b4ed738eee702b0e1d193c052d205/tox.ini#L174-L182 is of interest?

**Comment by H-Shay** [comment]
A few things: I am wondering if we can exclude the code that is specific to the c-implementation of frozendict from the coverage report. This code path is not touched at all unless the test runner is using the `frozendict` c-implementation, and the only way to ensure that is to ensure that the test runner (like a Github Actions runner) uses that specific implementation, which seems odd. This is the code that is causing the `coverage --fail-under-100` invocation to fail and blowing up the CI. 

Second, re: adding python 3.10-this blows up on the `nose` invocation, and it looks like `nose` is on it's way to being unmaintained: https://nose.readthedocs.io/en/latest/. In this case does it make sense to switch the dependency to something else, or to hold off on upgrading to python 3.10?

**Comment by DMRobertson** [comment]
> A few things: I am wondering if we can exclude the code that is specific to the c-implementation of frozendict from the coverage report. This code path is not touched at all unless the test runner is using the `frozendict` c-implementation, and the only way to ensure that is to ensure that the test runner (like a Github Actions runner) uses that specific implementation, which seems odd. This is the code that is causing the `coverage --fail-under-100` invocation to fail and blowing up the CI.

It'd be nice to have something that checks this works in both modes. But I'm not sure there's an easy way to explicitly say "please install frozendict with the C extension" and "please install without it". (I couldn't see anything in its [setup.py](https://github.com/Marco-Sulla/python-frozendict/blob/master/setup.py), anyway.) Given that, I think it's not worth the effort to account for both options here.

Suggestions:

- drop coverage arg `--fail-under-100`
- relax to `--fail-under=80` (for example)
- ignore as in 5c5f81c

> Second, re: adding python 3.10-this blows up on the `nose` invocation, and it looks like `nose` is on it's way to being unmaintained: https://nose.readthedocs.io/en/latest/. In this case does it make sense to switch the dependency to something else, or to hold off on upgrading to python 3.10?

Ouch---looks like something changed from deprecated to removed in 3.10 I think? That link suggests:

>  New projects should consider using Nose2, py.test, or just plain unittest/unittest2.

Maybe we can replace `nose` with `python -m unittest`? (If so, I'd suggest sanity checking that we're using the right Python version by e.g. calling `python --version`. I'm guessing that tox and the setup-python action will mean this just works, but... just a small bit of paranoia on my part there.)

If that doesn't work, I'd suggest giving nose2 a try.

**Comment by DMRobertson** [comment]
See https://github.com/H-Shay/python-canonicaljson/actions/runs/1309035566 for an example of this run in action

File: .github/workflows/tests.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+      - run: python -m pip install tox
+      - run: tox -e packaging
+
+  macos-tests:
+    runs-on: macos-latest
+    strategy:
+      matrix:
+        python-version: ["3.8", "3.9", "3.10"]
```
Comment (DMRobertson) [inline .github/workflows/tests.yaml:51]: I wasn't sure why we need to worry about macos specifically here. Looking at the history of the Travis CI config I found https://github.com/matrix-org/python-canonicaljson/commit/91009169bf6bed493d8c9f8360d1d3ec922fb045, which seems to suggest there were some string/byte encoding fun and games.

**Review by DMRobertson — APPROVED** [review]
LGTM!

**Pull request merged**